File: app/java/app/src/main/java/com/example/accounting/application/port/in/JournalUseCase.java
package com.example.accounting.application.port.in;

import com.example.accounting.domain.model.Journal;
import java.util.List;

/**
 * 仕訳ユースケース（Input Port）
 * ビジネスユースケースのインターフェース定義
 */
public interface JournalUseCase {

    /**
     * すべての仕訳を取得
     * @return 仕訳一覧
     */
    List<Journal> getAllJournals();

    /**
     * 仕訳番号で仕訳を取得
     * @param journalNo 仕訳番号
     * @return 仕訳
     */
    Journal getJournalByNo(String journalNo);

    /**
     * 仕訳を作成
     * @param journal 仕訳
     * @return 作成された仕訳
     */
    Journal createJournal(Journal journal);

    /**
     * 仕訳を更新
     * @param journalNo 仕訳番号
     * @param journal 仕訳
     * @return 更新された仕訳
     */
    Journal updateJournal(String journalNo, Journal journal);

    /**
     * 仕訳を削除
     * @param journalNo 仕訳番号
     */
    void deleteJournal(String journalNo);
}

File: app/scala/src/main/scala/com/example/accounting/application/port/in/JournalUseCase.scala
package com.example.accounting.application.port.in

import com.example.accounting.application.AppError
import com.example.accounting.domain.journal.*

import java.time.LocalDate

/**
 * 仕訳ユースケース（Input Port）
 */
trait JournalUseCase:

  /** 仕訳伝票番号で取得 */
  def getJournal(journalNo: String): Either[AppError, (Journal, List[JournalDetail], List[JournalDebitCreditDetail])]

  /** 日付範囲で仕訳を取得 */
  def getJournalsByDateRange(
      from: LocalDate,
      to: LocalDate,
  ): Either[AppError, List[(Journal, List[JournalDetail], List[JournalDebitCreditDetail])]]

  /** 仕訳を作成 */
  def createJournal(
      journal: Journal,
      details: List[JournalDetail],
      debitCreditDetails: List[JournalDebitCreditDetail],
  ): Either[AppError, (Journal, List[JournalDetail], List[JournalDebitCreditDetail])]

  /** 仕訳を削除 */
  def deleteJournal(journalNo: String): Either[AppError, Unit]

  /** 貸借平衡を検証 */
  def validateBalance(journalNo: String): Either[AppError, (Boolean, BigDecimal, BigDecimal)]
